AX: Vend AXSubrole=AXLandmarkForm for labeled form elements

rahimabdi opened the pull request:
#### a8728728ea0d85f52ec9162e6cee680f854fa5a7
<pre>
AX: Vend AXSubrole=AXLandmarkForm for labeled form elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=233907">https://bugs.webkit.org/show_bug.cgi?id=233907</a>
<a href="https://rdar.apple.com/86135614">rdar://86135614</a>

Reviewed by Tyler Wilcock.

Per ARIA specification, elements with a role of &quot;form&quot; (i.e., &lt;form&gt; or role=&quot;form&quot;) require an accessible name
and when not provided, this is an authoring error: <a href="https://w3c.github.io/aria/#document-handling_author-errors_roles.">https://w3c.github.io/aria/#document-handling_author-errors_roles.</a>
Form elements can be labelled like other landmarks (e.g., via aria-label, aria-labelledby) and additionally, native &lt;form&gt; elements
can be labeled with the title attribute.

This commit ensures that form elements are exposed as navigable landmarks when supplied with an accname. Form elements also now
vend AXSubrole=AXLandmarkForm to align with the classification of other landmarks.

* LayoutTests/accessibility/mac/search-predicate-expected.txt: Updated test expectation.
* LayoutTests/accessibility/mac/search-predicate.html: Updated &lt;form&gt; test.
* LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt Updated test expectation.
* LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt Updated test expectation.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isLandmark const):
(WebCore::AXCoreObject::ariaLandmarkRoleDescription const):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXARIAContentGroupText):

Canonical link: <a href="https://commits.webkit.org/294748@main">https://commits.webkit.org/294748@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/618fc3b9b592af554b313359b759cd26b072934b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/102938 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/22612 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/12932 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108106 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/53579 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/104977 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/22935 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31113 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/78255 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/35205 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/105944 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/17756 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/92879 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58590 "Passed tests") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-18-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/10925 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/52936 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/10992 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110479 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30075 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/22149 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87243 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30439 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89085 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/86866 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22110 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/31714 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9424 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/24337 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30002 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/35324 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/29810 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33137 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31372 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on LayoutTests/accessibility/gtk/aria-roles-unignored-expected.txt:
```diff
@@ -16,7 +16,7 @@ PASS element.role is 'AXRole: AXParagraph'
 PASS element.role is 'AXRole: AXLabel'
 PASS element.role is 'AXRole: AXHeading'
-PASS element.role is 'AXRole: AXForm'
+FAIL element.role should be AXRole: AXForm. Was AXRole: AXLandmarkForm.
```
Hmm, seems like this introduces a failure, which we don't want.

twilco commented on LayoutTests/accessibility/gtk/xml-roles-exposed-expected.txt:
```diff
@@ -6,8 +8,8 @@ On success, you will see a series of "PASS" messages, followed by "TEST COMPLETE
-AXPosition:  { 0.00000, 0.00000 }
-AXSize: { 0.00000, 0.00000 }
+AXPosition:  { 8.00000, 8.00000 }
+AXSize: { 784.000, 10.0000 }
```
Do you know why this change results in elements computing different geometry? That seems unexpected considering the intention of this change should just switch some roles around.

twilco commented on LayoutTests/accessibility/gtk/xml-roles-exposed-expected.txt:
```diff
@@ -19,14 +21,14 @@ AXMultiSelectable: 0
-AXChecked: 0q
-AXPlatformAttributes: atomic:true, relevant:additions text, live:assertive, container-live-role:alert, container-atomic:true, container-relevant:additions text, container-live:assertive, computed-role:alert, xml-roles:alert, tag:div, toolkit:WebKitGtk
+AXChecked: 0
+AXPlatformAttributes: atomic:true, computed-role:alert, container-atomic:true, container-live:assertive, container-live-role:alert, container-relevant:additions text, live:assertive, relevant:additions text, tag:div, xml-roles:alert
```
Do you know why your code change results in these attributes being rearranged? That also feels quite unexpected.

twilco commented on LayoutTests/accessibility/gtk/xml-roles-exposed-expected.txt:
```diff
@@ -115,13 +117,13 @@ AXEnabled: 1
 AXRequired: 0
 AXChecked: 0
-AXPlatformAttributes: computed-role:blockquote, xml-roles:blockquote, tag:div, toolkit:WebKitGtk
+AXPlatformAttributes: computed-role:blockquote, tag:div, xml-roles:blockquote
```
Why does your change result in `toolkit:WebKitGtk` being removed?

twilco commented on LayoutTests/accessibility/roles-exposed-expected.txt:
```diff
@@ -1,385 +1,1417 @@
+      Could not get AX element.
+
+abbr[title]
+      Could not get AX element.
```
The diff on this file is huge, and most of the changes seem unrelated to forms, which is pretty unexpected. Why has this changed so much?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2821,8 +2821,15 @@ AccessibilityRole AccessibilityNodeObject::determineAriaRoleAttribute() const
+            "require an accessible name, such as a non-empty aria-label attribute, to ensure the region is a navigable landmark. For more details, "
+            "see the ARIA specification: https://w3c.github.io/aria/#document-handling_author-errors_roles."_s);
+        auto* document = this->document();
+        document->addConsoleMessage(MessageSource::Accessibility, MessageLevel::Warning, consoleMessage);
```
We should null-check document before using it to avoid crashes.

twilco commented on Source/WebCore/accessibility/atspi/AccessibilityObjectAtspi.cpp:
```diff
@@ -288,8 +288,6 @@ static Atspi::Role atspiRole(AccessibilityRole role)
     case AccessibilityRole::Footer:
         return Atspi::Role::Footer;
-    case AccessibilityRole::Form:
-        return Atspi::Role::Form;
```
This feels like a pretty significant change for the ATSPI platform. Are we worried about ATs on that platform no longer announcing form, no longer seeing form containers and thus losing important semantics, etc?

twilco commented on Source/WebCore/accessibility/AccessibilityNodeObject.cpp:
```diff
@@ -2835,7 +2835,7 @@ AccessibilityRole AccessibilityNodeObject::determineAriaRoleAttribute() const
+        // If a form or region has no label, but it does have a fallback role, use that instead.
         auto nextRole = ariaRoleToWebCoreRole(ariaRole, [] (const AccessibilityRole& skipRole) {
             return skipRole == AccessibilityRole::LandmarkRegion;
         });
```
I think in order for this comment change to be accurate, we'd need to change the lambda too:

`return skipRole == AccessibilityRole::LandmarkRegion || AccessibilityRole::Form;`

But then we'd probably need a new WPT test case verifying this behavior too (unless there already is one — I know there is for the region role, which is why this code was added), and I'm guessing you probably want to move forward from this change 🙂 

Could be a good follow-up? If you agree, let's remove the change in the comment for now.

rahimabdi replied:
Agreed! 

Removed this change and filed an a11y interop issue for testing labeled `<form>` elements being exposed as `role="form"`, and role fallback when unlabeled: https://github.com/web-platform-tests/interop-accessibility/issues/188.

twilco reviewed (commented):
Very nice! Just one last comment.

rahimabdi committed:
AX: Vend AXSubrole=AXLandmarkForm for labeled form elements
https://bugs.webkit.org/show_bug.cgi?id=233907
rdar://86135614

Reviewed by Tyler Wilcock.

Per ARIA specification, elements with a role of "form" (i.e., <form> or role="form") require an accessible name
and when not provided, this is an authoring error: https://w3c.github.io/aria/#document-handling_author-errors_roles.
Form elements can be labelled like other landmarks (e.g., via aria-label, aria-labelledby) and additionally, native <form> elements
can be labeled with the title attribute.

This commit ensures that form elements are exposed as navigable landmarks when supplied with an accname. Form elements also now
vend AXSubrole=AXLandmarkForm to align with the classification of other landmarks.

* LayoutTests/accessibility/mac/search-predicate-expected.txt: Updated test expectation.
* LayoutTests/accessibility/mac/search-predicate.html: Updated <form> test.
* LayoutTests/platform/mac-wk1/accessibility/roles-exposed-expected.txt Updated test expectation.
* LayoutTests/platform/mac-wk2/accessibility/roles-exposed-expected.txt Updated test expectation.
* Source/WebCore/accessibility/AXCoreObject.cpp:
(WebCore::AXCoreObject::isLandmark const):
(WebCore::AXCoreObject::ariaLandmarkRoleDescription const):
* Source/WebCore/accessibility/mac/AccessibilityObjectMac.mm:
(WebCore::AccessibilityObject::subrolePlatformString const):
* Source/WebCore/accessibility/mac/CocoaAccessibilityConstants.h:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::AXARIAContentGroupText):

Canonical link: https://commits.webkit.org/294748@main
